core/storage/keyvalue.hpp

```diff
@@ -0,0 +1,26 @@
+#ifndef KAGOME_KEYVALUE_HPP
+#define KAGOME_KEYVALUE_HPP
+
+namespace kagome::storage {
+
+class KeyValue {
+ public:
+
+  /**
+   * @brief Store value by key
+   * @param key non-empty byte buffer
+   * @param value byte buffer
+   */
+  virtual void put(const Buffer &key, const Buffer &value) = 0;
+
+  /**
+   * @brief Get value by key
+   * @param key non-empty byte buffer
+   * @return buffer of size 0 if no data is stored given {@param key} or buffer with data
+   */
+  virtual Buffer get(const Buffer &key) const = 0;
```

Comment:
Isn't it possible to avoid Buffer copy? For example, by returning a constref?

Reply:
implementation will be reading data and allocating it for the consumption by client classes. *Where* should be stored that reference? You can't return ref to local copy

Reply:
RVO will work here, so it is fine to leave it as is